pandas/core/index.py

```diff
@@ -3818,6 +3859,544 @@ def _wrap_joined_index(self, joined, other):
+
+            # seems we only have indexing ops to infer
+            # rather than direct accessors
+            if len(start) > 1:
+                step = start[1] - start[0]
+                stop = start[-1] + step
+                start = start[0]
+            elif len(start):
+                start = start[0]
+                stop = start + 1
+                step = 1
+            else:
+                start = stop = 0
+                step = 1
+
+        # pass thru to main constructor
+        elif isinstance(start, (Index, np.ndarray, ABCSeries)):
+            return Index(start, dtype=stop, copy=step, name=name, **kwargs)
+
+        # validate the arguments
+        def _ensure_int(value, field):
+            try:
+                new_value = int(value)
+            except:
+                new_value = value
+
+            if not is_integer(new_value) or new_value != value:
+                raise TypeError("RangeIndex(...) must be called with integers,"
+                                " {value} was passed for {field}".format(
+                                    value=value,
+                                    field=field)
+                                )
+
+            return new_value
+
+        if start is None:
+            start = 0
+        else:
+            start = _ensure_int(start, 'start')
+        if stop is None:
+            stop = start
+            start = 0
+        else:
+            stop = _ensure_int(stop, 'stop')
+        if step is None:
+            step = 1
+        elif step == 0:
+            raise ValueError("Step must not be zero")
+        else:
+            step = _ensure_int(step, 'step')
+
+        return cls._simple_new(start, stop, step, name)
+
+    @classmethod
+    def _simple_new(cls, start, stop=None, step=None, name=None,
+                    dtype=None, **kwargs):
+        result = object.__new__(cls)
+
+        # handle passed None, non-integers
+        if start is None or not is_integer(start):
+            return RangeIndex(start, stop, step, name=name, **kwargs)
+
+        result._start = start
+        result._stop = stop or 0
+        result._step = step or 1
+        result.name = name
+        for k, v in compat.iteritems(kwargs):
+            setattr(result, k, v)
+
+        result._reset_identity()
+        return result
+
+    @staticmethod
+    def _validate_dtype(dtype):
+        """ require dtype to be None or int64 """
+        if not (dtype is None or is_int64_dtype(dtype)):
+            raise TypeError('Invalid to pass a non-int64 dtype to RangeIndex')
+
+    @cache_readonly
+    def _data(self):
+        return np.arange(self.start, self.stop, self.step, dtype=np.int64)
+
+    @cache_readonly
+    def _int64index(self):
+        return Int64Index(self._data, name=self.name, fastpath=True)
+
+    def __reduce__(self):
+        d = self._get_attributes_dict()
+        d['start'] = self.start
+        d['stop'] = self.stop
+        d['step'] = self.step
```

Comment:
Ignore if you've thought about this, but why not have start, stop & step as attributes, and then `d` will already contain them? (and you might not need this or the next method at all)